Adding OS version info to nodes' `Info` struct and to the system info's API

wk8 opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

This patch adds the OS' version to nodes' `Info` struct and to the system info's API.

This is needed so that we can add OS version constraints in Swarmkit, which
does require the engine to report its host's OS version (see
https://github.com/docker/swarmkit/issues/2770#issuecomment-437211288).

**- How I did it**

The OS version is parsed from the `os-release` file on Linux, and from the
`ReleaseId` string value of the `SOFTWARE\Microsoft\Windows NT\CurrentVersion`
registry key on Windows.

**- How to verify it**

Unit test included.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

Adding OS version info to the system info's API (`/info`)

**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah reviewed (changes requested):
I'm a bit concerned about using this for scheduling as we'll now end up with two vectors to schedule (image manifest / platform as advertised by engine info), which may not provide the same information (e.g. Architecture may also have to be taken into account, as well as isolation (hyper-v / process isolation).

In addition; we should consider having a different way to obtain this info, because the `/info` endpoint is far from lightweight to call

wk8 commented on pkg/parsers/operatingsystem/operatingsystem_windows.go:
```diff
@@ -8,40 +8,53 @@ import (
+}
+
+func withCurrentVersionRegistryKey(f func(registry.Key) (string, error)) (string, error) {
+	key, err := registry.OpenKey(registry.LOCAL_MACHINE, `SOFTWARE\Microsoft\Windows NT\CurrentVersion`, registry.QUERY_VALUE)
```
This one gives the major/minor/patch version of Windows (eg 6.2.9200); whereas as far as I can tell Windows folks tend to prefer using the release ID to refer to Windows Server versions (see eg https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/system-requirements#virtualized-container-hosts or any of our own doc around Windows.

But I might be totally wrong on this, and I'm happy to change to whichever Windows-savvy people think makes more sense; since this patch is aimed primarily at UCP folks anyway, maybe @wsong or @alexmavr can give some insight here as to which version number is relevant here? Or maybe @dhiltgen ?

wk8 commented:
Thanks for the quick review, @thaJeztah , much appreciated :)

> In addition; we should consider having a different way to obtain this info, because the /info endpoint is far from lightweight to call
That endpoint seems to be the one, central, place where we gather info on hosts (and is tied directly to the `types.Info` struct, that is also the one, central place for this); so I'm not sure where else it would make sense to do this; could you please elaborate a bit on where you'd rather have this? Or are you suggesting we should start breaking up `types.Info` into multiple structs?

> I'm a bit concerned about using this for scheduling as we'll now end up with two vectors to schedule (image manifest / platform as advertised by engine info), which may not provide the same information (e.g. Architecture may also have to be taken into account, as well as isolation (hyper-v / process isolation).
Agreed that there is food for thought here, as there is a lot of info we could use to match images & nodes that we don't really make use of just yet (we pretty much only use the architecture for now); however, that seems to be a discussion more relevant on the upcoming swarmkit change. This patch here is only concerned with providing more info to swarmkit.
We've been discussing that scheduling change with @wsong @anshulpundir  and @dhiltgen , notably on https://github.com/docker/swarmkit/issues/2770, and this is the best we've come up with. Happy to brain storm more with with you any time though :)

thaJeztah commented:
> That endpoint seems to be the one, central, place where we gather info on hosts (and is tied directly to the types.Info struct, that is also the one, central place for this)

Yes, and I think that's a problem, because the endpoint just about touches every subsystem of dockerd (gathering information about containers (and their state), images, which plugins are available, low-level information from the storage-driver). Some of those have caused issues in the past (e.g. due to locks when determining container state).

I think we need a way to get more fine-grained information out of the daemon, especially if this endpoint is gonna be used more.

> Agreed that there is food for thought here, as there is a lot of info we could use to match images & nodes that we don't really make use of just yet

I think the key discussion here is (also see https://github.com/moby/moby/issues/34875) that we need to finish the standardisation of the "platform" information on images, and set the same information on the daemon. I think the last discussion ended with;

- review/define the fields that are included in the platform
- make sure that there's a well-defined format to serialize the platform in a string
- the platform (and string representation) should allow _optional_ fields (e.g. OS version, and OS features may not be present on all platforms)
- when parsing the string, those fields can be ignored by a scheduler (and when matching (multi-arch) images with the daemon/runtime)



Related issues;

- https://github.com/moby/moby/pull/35090 Add support for Windows version filtering on pull
- https://github.com/containerd/containerd/issues/2293 Support running ARM images without variant information
- https://github.com/containerd/containerd/issues/2133 ARM Variant Matching
- https://github.com/containerd/containerd/issues/2029 Discuss: How to handle multi-arch images when platforms support multi-arch runtimes?
- (slightly related https://github.com/containerd/containerd/issues/2594 Utilize `os.features` for plugin images with linux library requirements)


There was a lengthy discussion on the original PR that added the OS and Architecture; https://github.com/moby/moby/pull/13921


> This patch here is only concerned with providing more info to swarmkit.

Correct; it's probably fine to show this information for _informational purposes_; in which case the field should carry a big disclaimer that it's meant as-such, and that its content (or format) is not part of the API specification. Similar to what we added to the `DriverStatus` field; https://github.com/moby/moby/blob/bb7de1f7cbfa196b0972a3f8e3ac2a554c5942cc/api/swagger.yaml#L3579-L3592

however (glancing over the discussion on the swarmkit issue), IIUC, the purpose of this change is to influence scheduling decisions (manually or automatically). If the information used here does not match the direction that image-selection will use (the discussion on https://github.com/moby/moby/issues/34875), there is a risk;

- user specifies constraint to run service on `Windows version Foobar.X.Y.Z` nodes
- service fails to deploy, because manifest selection uses a different algorithm
- or service fails to deploy because the image is not compatible
- confusion..


Note that for _scheduling_, another approach would be to set up engine (or node-) labels, and set constraints based on those labels.

thaJeztah commented on pkg/parsers/operatingsystem/operatingsystem_linux_test.go:
```diff
@@ -38,10 +43,7 @@ PRETTY_NAME=Ubuntu 14.04.LTS`,
 		},
 	}
 
-	valids := []struct {
```
Instead of having `valid` and `invalid` types, perhaps just combine them in a single list. I gave this a quick spin (also turning them into subtests), but let me know what you think;

```patch
diff --git a/pkg/parsers/operatingsystem/operatingsystem_linux_test.go b/pkg/parsers/operatingsystem/operatingsystem_linux_test.go
index 2eb9d47673..5d35cd4024 100644
--- a/pkg/parsers/operatingsystem/operatingsystem_linux_test.go
+++ b/pkg/parsers/operatingsystem/operatingsystem_linux_test.go
@@ -7,45 +7,39 @@ import (
 	"os"
 	"path/filepath"
 	"testing"
-)
-
-type valid struct {
-	content  string
-	expected string
-}
 
-type invalid struct {
-	content       string
-	expectedError string
-}
+	"gotest.tools/assert"
+)
 
 func TestGetOperatingSystem(t *testing.T) {
-	invalids := []invalid{
+	tests := []struct {
+		name        string
+		content     string
+		expected    string
+		expectedErr string
+	}{
 		{
-			`PRETTY_NAME=Source Mage GNU/Linux
+			content: `PRETTY_NAME=Source Mage GNU/Linux
 PRETTY_NAME=Ubuntu 14.04.LTS`,
-			"PRETTY_NAME needs to be enclosed by quotes if they have spaces: Source Mage GNU/Linux",
+			expectedErr: "PRETTY_NAME needs to be enclosed by quotes if they have spaces: Source Mage GNU/Linux",
 		},
 		{
-			`PRETTY_NAME="Ubuntu Linux
+			content: `PRETTY_NAME="Ubuntu Linux
 PRETTY_NAME=Ubuntu 14.04.LTS`,
-			"PRETTY_NAME is invalid: invalid command line string",
+			expectedErr: "PRETTY_NAME is invalid: invalid command line string",
 		},
 		{
-			`PRETTY_NAME=Ubuntu'
+			content: `PRETTY_NAME=Ubuntu'
 PRETTY_NAME=Ubuntu 14.04.LTS`,
-			"PRETTY_NAME is invalid: invalid command line string",
+			expectedErr: "PRETTY_NAME is invalid: invalid command line string",
 		},
 		{
-			`PRETTY_NAME'
+			content: `PRETTY_NAME'
 PRETTY_NAME=Ubuntu 14.04.LTS`,
-			"PRETTY_NAME needs to be enclosed by quotes if they have spaces: Ubuntu 14.04.LTS",
+			expectedErr: "PRETTY_NAME needs to be enclosed by quotes if they have spaces: Ubuntu 14.04.LTS",
 		},
-	}
-
-	valids := []valid{
 		{
-			`NAME="Ubuntu"
+			content: `NAME="Ubuntu"
 PRETTY_NAME_AGAIN="Ubuntu 14.04.LTS"
 VERSION="14.04, Trusty Tahr"
 ID=ubuntu
@@ -54,10 +48,10 @@ VERSION_ID="14.04"
 HOME_URL="http://www.ubuntu.com/"
 SUPPORT_URL="http://help.ubuntu.com/"
 BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"`,
-			"Linux",
+			expected: "Linux",
 		},
 		{
-			`NAME="Ubuntu"
+			content: `NAME="Ubuntu"
 VERSION="14.04, Trusty Tahr"
 ID=ubuntu
 ID_LIKE=debian
@@ -65,10 +59,10 @@ VERSION_ID="14.04"
 HOME_URL="http://www.ubuntu.com/"
 SUPPORT_URL="http://help.ubuntu.com/"
 BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"`,
-			"Linux",
+			expected: "Linux",
 		},
 		{
-			`NAME=Gentoo
+			content: `NAME=Gentoo
 ID=gentoo
 PRETTY_NAME="Gentoo/Linux"
 ANSI_COLOR="1;32"
@@ -76,10 +70,10 @@ HOME_URL="http://www.gentoo.org/"
 SUPPORT_URL="http://www.gentoo.org/main/en/support.xml"
 BUG_REPORT_URL="https://bugs.gentoo.org/"
 `,
-			"Gentoo/Linux",
+			expected: "Gentoo/Linux",
 		},
 		{
-			`NAME="Ubuntu"
+			content: `NAME="Ubuntu"
 VERSION="14.04, Trusty Tahr"
 ID=ubuntu
 ID_LIKE=debian
@@ -88,48 +82,62 @@ VERSION_ID="14.04"
 HOME_URL="http://www.ubuntu.com/"
 SUPPORT_URL="http://help.ubuntu.com/"
 BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"`,
-			"Ubuntu 14.04 LTS",
+			expected: "Ubuntu 14.04 LTS",
 		},
 		{
-			`NAME="Ubuntu"
+			content: `NAME="Ubuntu"
 VERSION="14.04, Trusty Tahr"
 ID=ubuntu
 ID_LIKE=debian
 PRETTY_NAME='Ubuntu 14.04 LTS'`,
-			"Ubuntu 14.04 LTS",
+			expected: "Ubuntu 14.04 LTS",
 		},
 		{
-			`PRETTY_NAME=Source
+			content: `PRETTY_NAME=Source
 NAME="Source Mage"`,
-			"Source",
+			expected: "Source",
 		},
 		{
-			`PRETTY_NAME=Source
+			content: `PRETTY_NAME=Source
 PRETTY_NAME="Source Mage"`,
-			"Source Mage",
+			expected: "Source Mage",
 		},
 	}
 
-	testParsingOsRelease(t, GetOperatingSystem, invalids, valids)
+	defer setupTest()()
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if err := ioutil.WriteFile(etcOsRelease, []byte(test.content), 0600); err != nil {
+				t.Fatalf("failed to write to %s: %v", etcOsRelease, err)
+			}
+			s, err := GetOperatingSystem()
+			if test.expectedErr == "" {
+				assert.NilError(t, err)
+			} else {
+				assert.Error(t, err, test.expectedErr)
+			}
+			assert.Equal(t, s, test.expected)
+		})
+	}
 }
 
 func TestGetOperatingSystemVersion(t *testing.T) {
-	invalids := []invalid{
+	tests := []struct {
+		name        string
+		content     string
+		expected    string
+		expectedErr string
+	}{
 		{
-			`VERSION_ID="18.04
+			name: "invalid version id",
+			content: `VERSION_ID="18.04
 VERSION_ID=18.04`,
-			"VERSION_ID is invalid: invalid command line string",
+			expectedErr: "VERSION_ID is invalid: invalid command line string",
 		},
 		{
-			`VERSION_ID=18.04'
-VERSION_ID=18.04`,
-			"VERSION_ID is invalid: invalid command line string",
-		},
-	}
-
-	valids := []valid{
-		{
-			`NAME="Ubuntu"
+			name: "ubuntu 14.04",
+			content: `NAME="Ubuntu"
 PRETTY_NAME="Ubuntu 14.04.LTS"
 VERSION="14.04, Trusty Tahr"
 ID=ubuntu
@@ -138,10 +146,11 @@ VERSION_ID="14.04"
 HOME_URL="http://www.ubuntu.com/"
 SUPPORT_URL="http://help.ubuntu.com/"
 BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"`,
-			"14.04",
+			expected: "14.04",
 		},
 		{
-			`NAME=Gentoo
+			name: "gentoo",
+			content: `NAME=Gentoo
 ID=gentoo
 PRETTY_NAME="Gentoo/Linux"
 ANSI_COLOR="1;32"
@@ -149,47 +158,42 @@ HOME_URL="http://www.gentoo.org/"
 SUPPORT_URL="http://www.gentoo.org/main/en/support.xml"
 BUG_REPORT_URL="https://bugs.gentoo.org/"
 `,
-			"",
 		},
 		{
-			`VERSION_ID="14.04"
+			name: "dual version id",
+			content: `VERSION_ID="14.04"
 VERSION_ID=18.04`,
-			"18.04",
+			expected: "18.04",
 		},
 	}
 
-	testParsingOsRelease(t, GetOperatingSystemVersion, invalids, valids)
+	defer setupTest()()
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if err := ioutil.WriteFile(etcOsRelease, []byte(test.content), 0600); err != nil {
+				t.Fatalf("failed to write to %s: %v", etcOsRelease, err)
+			}
+			s, err := GetOperatingSystemVersion()
+			if test.expectedErr == "" {
+				assert.NilError(t, err)
+			} else {
+				assert.Error(t, err, test.expectedErr)
+			}
+			assert.Equal(t, s, test.expected)
+		})
+	}
 }
 
-func testParsingOsRelease(t *testing.T, funcToTest func() (string, error), invalids []invalid, valids []valid) {
+func setupTest() func() {
 	var backup = etcOsRelease
 
 	dir := os.TempDir()
 	etcOsRelease = filepath.Join(dir, "etcOsRelease")
 
-	defer func() {
+	return func() {
 		os.Remove(etcOsRelease)
 		etcOsRelease = backup
-	}()
-
-	for _, elt := range invalids {
-		if err := ioutil.WriteFile(etcOsRelease, []byte(elt.content), 0600); err != nil {
-			t.Fatalf("failed to write to %s: %v", etcOsRelease, err)
-		}
-		s, err := funcToTest()
-		if err == nil || err.Error() != elt.expectedError {
-			t.Errorf("Expected an error %q, got %q (err: %v)", elt.expectedError, s, err)
-		}
-	}
-
-	for _, elt := range valids {
-		if err := ioutil.WriteFile(etcOsRelease, []byte(elt.content), 0600); err != nil {
-			t.Fatalf("failed to write to %s: %v", etcOsRelease, err)
-		}
-		s, err := funcToTest()
-		if err != nil || s != elt.expected {
-			t.Errorf("Expected %q, got %q (err: %v)", elt.expected, s, err)
-		}
 	}
 }
 ```

wk8 replied:
Thanks, I like it. DRY-ed up just a bit further.

wk8 commented on pkg/parsers/operatingsystem/operatingsystem_linux_test.go:
```diff
@@ -108,6 +110,60 @@ PRETTY_NAME="Source Mage"`,
+		},
+		{
+			`VERSION_ID=18.04'
+VERSION_ID=18.04`,
```
Indeed, changed

olljanat commented:
Derek add label: rebuild/*

thaJeztah commented on daemon/info.go:
```diff
@@ -251,6 +252,14 @@ func operatingSystem() string {
 	return operatingSystem
 }
 
+func osVersion() (version string) {
```
Discussing with @cpuguy83 - we should probably store the value, so that we don't have to actually fetch it each time we call the `/info` endpoint (in order to make it not more heavy)

Also; we might want to add it to the prometheus metrics

thaJeztah commented:
Discussing if we can come up with alternatives for this use-case, but will have to do a bit more thinking about that

thaJeztah commented:
Discussing in the maintainers meeting (again 😅 );

- can you add this to the prometheus metrics?
- can you cache the results?

Other than that, we're ok with this change

wk8 commented:
Great, thanks @thaJeztah ! :)

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "wk8/os_version" git@github.com:wk8/moby.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354462472
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

wk8 commented:
@thaJeztah : amended as requested, could you please have another look? Thanks!

kolyshkin commented on api/swagger.yaml:
```diff
@@ -3758,6 +3758,17 @@ definitions:
+
+          > **Note**: The information returned in this field, including the
+          > its existence, and the formatting of values, should not be considered,
+          > stable, and may change without notice.
```
"including the its existence" does not sound right to me

"considered, stable" -- looks like comma should not be here

wk8 replied:
Good catch, fixed!

thaJeztah commented on daemon/info.go:
```diff
@@ -233,23 +239,48 @@ func memInfo() *system.MemInfo {
 }
 
+// this caches the result of operatingSystem below, for efficiency
+var operatingSystemResult string
```
@cpuguy83 think we need a lock for this variable?

thaJeztah commented on daemon/info.go:
```diff
@@ -233,23 +239,48 @@ func memInfo() *system.MemInfo {
+	defer metrics.StartTimer(hostInfoFunctions.WithValues("operating_system"))()
+
+	if operatingSystemResult == "" {
+		var operatingSystem string
```
nit; wondering if we should just move this variable out instead of introducing the `operatingSystemResult` variable (definitely open to suggestions there)

thaJeztah commented on daemon/info.go:
```diff
@@ -233,23 +239,48 @@ func memInfo() *system.MemInfo {
+func osVersion() string {
+	defer metrics.StartTimer(hostInfoFunctions.WithValues("os_version"))()
+
+	if osVersionResult == "" {
```
Same here; could use an early return;

```suggestion
	if osVersionResult != "" {
		return osVersionResult
	}
```

thaJeztah reviewed (changes requested):
Some quick comments 😅

wk8 commented:
@thaJeztah amended as requested. Could you please take another look? Thanks!

thaJeztah reviewed (changes requested):
two questions (also had the remaining one about possibly needing a lock, but I'll discuss that one later)

wk8 commented on daemon/info.go:
```diff
@@ -252,7 +266,29 @@ func operatingSystem() string {
+
+	operatingSystemResult = &operatingSystem
+
+	return
```
👍

wk8 commented on daemon/info.go:
```diff
@@ -252,7 +266,29 @@ func operatingSystem() string {
+	}
+	osVersionResult = &version
+
+	return
```
👍

wk8 commented:
@thaJeztah amended as requested. Could you please take another look? Thanks!

thaJeztah reviewed (approved):
LGTM

can you squash the two commits?

wk8 commented:
@thaJeztah : squashed as requested

thaJeztah commented:
ping @cpuguy83 PTAL

cpuguy83 commented on pkg/parsers/operatingsystem/operatingsystem_windows.go:
```diff
@@ -3,45 +3,57 @@ package operatingsystem // import "github.com/docker/docker/pkg/parsers/operatin
-	if err != nil {
-		return ret, err
+	os, err := withCurrentVersionRegistryKey(func(key registry.Key) (os string, err error) {
+		if os, _, err = key.GetStringValue("ProductName"); err != nil {
```
@jhowardmsft All this LGTY?

jhowardmsft replied:
Changes look fine. Just not sure if you we want "just" 10.0.17763 being returned in the API (on Windows), or something fuller (possibly as well) which also includes the SKU/UBR etc.

cpuguy83 commented on daemon/info.go:
```diff
@@ -236,8 +242,19 @@ func memInfo() *system.MemInfo {
+var operatingSystemResult *string
+
+func operatingSystem() (operatingSystem string) {
+	if operatingSystemResult != nil {
```
This does not look safe.
There can be multple calls here and end up with two goroutines trying to set the same value and corrupting memory.

wk8 replied:
@cpuguy83 : that's addressed by the comment just above (lines 246-248). @thaJeztah seemed to agree there. Happy to discuss more if you think that comment is erroneous.

cpuguy83 replied:
Hmm... I don't think this comment is true.

cpuguy83 replied:
AFAIK the spec does not say that setting pointers is atomic even if it may be in practice, so we cannot rely on this.
But if we set the value in one place at daemon startup, this should be fine.

wk8 replied:
Changed to using a mutex.

kolyshkin replied:
Maybe the code will look easier if `sync.Once` was used?

cpuguy83 reviewed (changes requested):
I think if we are going to memoize the operating system result we need something to synchronize this.
I think, however, this could just be set on startup.
I seem to recall we talked about adding the operating system value to the metrics details like we have other system info, which is initialized here: https://github.com/moby/moby/blob/54dddadc7d5d89fe0be88f76979f6f6ab0dede83/daemon/daemon.go#L1049-L1058

wk8 commented:
@cpuguy83 : answered the question about safety. As to moving it to https://github.com/moby/moby/blob/54dddadc7d5d89fe0be88f76979f6f6ab0dede83/daemon/daemon.go#L1049-L1058, not opposed to doing that if you feel strongly about this; one argument against it could be that this is seen as a temporary change until we come up with a robust way of specifying OS flavours and versions for manifest files and re-use that here.

wk8 commented:
@cpuguy83 : amended as requested, PTAL, thanks!

cpuguy83 commented:
Can we just store the value on startup instead not knowing if it's been set or not?

StefanScherer commented:
Thanks @wk8 for this PR. @thaJeztah guided me to your PR as I have another use-case to get the Windows operating system version.

I did some brainstorming about Windows Image Tag names yesterday with Markos from Microsoft yesterday. The new Windows base images are also tagged with eg. "10.0.17763.379".

I find this tag format useful as I normally look into the manifest lists to see if an image supports Windows or not and which OS version of Windows is supported. There's a tool from Phil Estes to have a good overview:

```
$ docker run --rm mplatform/mquery golang
Image: golang
 * Manifest List: Yes
 * Supported platforms:
   - linux/amd64
   - linux/arm/v7
   - linux/arm64
   - linux/386
   - linux/ppc64le
   - linux/s390x
   - windows/amd64:10.0.14393.2791
   - windows/amd64:10.0.16299.967
   - windows/amd64:10.0.17134.590
   - windows/amd64:10.0.17763.316
```

A `docker info` retrieves all four numbers for this tag, but we don't get it in one piece that could be used.

```
$ docker info --format "{{ .KernelVersion }}"
10.0 17763 (17763.1.amd64fre.rs5_release.180914-1434)
$ docker info --format "{{ .OperatingSystem }}"
Windows Server 2019 Datacenter Version 1809 (OS Build 17763.379)
```

When we inspect images we see there is a metadata `OsVersion`:

```
$ docker image inspect mcr.microsoft.com/windows/nanoserver:10.0.17763.379 | jq '.[].OsVersion'
"10.0.17763.379"
```

Now when `docker info` could give us this string `10.0.17763.379` in one piece it would be easy for users to fetch the best matching base image for the current operating system:

I thought about changing the `KernelVersion` to that format.

Even in the Windows Server Insider preview it would be easy to fetch the matching base image with

```
$ docker image pull mcr.microsoft.com/windows/nanoserver:$(docker info --format "{{ .KernelVersion }}")
```

Another idea is adding the `OsVersion` to the info response and keep the KernelVersion format as it is. Then it would be `docker info --format "{{ .OsVersion }}`


I see your PR fills the first three numbers. For the image tagging we would need all four numbers in it.

Any opinions/ideas/feedback? :slightly_smiling_face:

thaJeztah commented:
> Any opinions/ideas/feedback? 🙂

There's (unfortunately) quite some "human friendly presentation" in our API responses; If the API returned the pieces as ("strict") elements, we could add a template function in the CLI to combine them, so that it could be used in a Go template;

```bash
docker info --format '{{.GetMeAllTheThings}}'

# equivalent of
docker info --format '{{.Kernel.Major}}.{{.Kernel.Minor}}.{{.Kernel.Build}}'
```

Not sure what makes sense though (i.e. should we that specific in the API and/or what to put there for other platforms?)

david-yu commented:
@wk8 Could you revisit this PR and address @cpuguy83's comments. Thanks in advance!

wk8 commented:
@david-yu will do next week, thanks :)

thaJeztah commented:
@wk8 was this updated?

wk8 committed:
Adding OS version info to the nodes' `Info` struct

This is needed so that we can add OS version constraints in Swarmkit, which
does require the engine to report its host's OS version (see
https://github.com/docker/swarmkit/issues/2770).

The OS version is parsed from the `os-release` file on Linux, and from the
`ReleaseId` string value of the `SOFTWARE\Microsoft\Windows NT\CurrentVersion`
registry key on Windows.

Added unit tests when possible, as well as Prometheus metrics.

Signed-off-by: Jean Rouge <rougej+github@gmail.com>

wk8 commented:
@thaJeztah : amended as requested per @cpuguy83 . Thanks.

cpuguy83 reviewed (approved):
LGTM
